cmd_fetchurl.go

```diff
@@ -0,0 +1,102 @@
+package main
+
+import (
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"path/filepath"
+
+	"github.com/direnv/direnv/sri"
+)
+
+// CmdFetchURL is `direnv fetchurl <URL> [SRI_HASH]`
+var CmdFetchURL = &Cmd{
+	Name:   "fetchurl",
+	Desc:   "Fetches a given URL",
+	Args:   []string{"URL", "SRI_HASH"},
+	Action: actionWithConfig(cmdFetchURL),
+}
+
+func cmdFetchURL(env Env, args []string, config *Config) (err error) {
+	if len(args) < 2 {
+		return fmt.Errorf("missing URL argument")
+	}
+
+	var (
+		algo sri.Algo = sri.SHA256
+		url string
+		sriHash string
+		cacheFile string
+	)
+
+	url = args[1]
+	// validate the SRI hash if it exists
+	if len(args) > 2 {
+		sriHash = args[2]
+		algo, err = sri.GetAlgo(sriHash)
+		if err != nil {
+			return err
+		}
+
+		// shortcut if the cache already has the file
+		cacheFile = filepath.Join(config.CacheDir, "fetchurl", sriHash)
+		if fileExists(cacheFile) {
+			fmt.Println(cacheFile)
+			return nil
+		}
+	}
+
+	tmpfile, err := ioutil.TempFile("", "fetchurl")
+	if err != nil {
+		return err
+	}
+	defer os.Remove(tmpfile.Name()) // clean up
+
+	resp, err := http.Get(url)
+	if err != nil {
+		return err
+	}
+
+	w := sri.NewWriter(tmpfile, algo)
+	if _, err = io.Copy(w, resp.Body); err != nil {
+		return err
+	}
+
+	newSRIHash := w.Sum()
+
+	// Validate
+	if sriHash != "" && newSRIHash != sriHash {
+		return fmt.Errorf("hash mismatch. Expected '%s' but got '%s'", sriHash, newSRIHash)
+	}
+
+	cacheFile = filepath.Join(config.CacheDir, "fetchurl", newSRIHash)
+	if fileExists(cacheFile) {
+		fmt.Println(cacheFile)
+		return nil
+	}
+
+	if err = os.MkdirAll(filepath.Dir(cacheFile), os.FileMode(0755)); err != nil {
+		return err
+	}
+
+	f, err := os.Create(cacheFile)
+	if err != nil {
+		return err
+	}
+	r, err := os.Open(tmpfile.Name())
+	if err != nil {
+		return err
+	}
+	if _, err = io.Copy(f, r); err != nil {
+		return err
+	}
+
+	// Make the file read-only and executable
+	if err = f.Chmod(os.FileMode(0500)); err != nil {
+		return err
+	}
+	
+	_, err = fmt.Println(cacheFile)
```

Comment:
One could check if stdout is a pipe or a tty to decide wether to print the instructions on stdout or stderr.